# Leren: Programming assignment 4

**Student 1:**  <span style="color:red">Tycho Koster</span> (<span style="color:red">10667687</span>)<br>
**Student 2:** <span style="color:red">David Stap</span> (<span style="color:red">10608516</span>)<br>

### Read logistic data

In [2]:
import operator
import math
import numpy as np
from numpy import genfromtxt


# import X and y from csv file
def read_logistic_data(filename):
    my_data = genfromtxt(filename, delimiter=';')
    X = []
    y= []
    for i in range(len(my_data)):
        new_data = my_data[i]
        y.append([new_data[-1]])
        X.append(np.delete(new_data, -1))
    return X, y

X_train, y_train = read_logistic_data('digits123-1.csv')
X_test, y_test = read_logistic_data('digits123-2.csv')

### Question 1): k NN
For both unweighted k-NN and weighted k-NN the same function is used. User can choose between these options by giving argument weight=False for unweighted and weight=True for weighted.

In [3]:
'''
INPUT: 
- n-dimensional vector x (numpy array)
- n-dimensional vector y (numpy array)

OUTPUT: 
- euclidean distance of vectors x and y
'''
def distance(x, y):
    return math.sqrt(sum((x-y)**2))

'''
INPUT:
- training set X (training examples with features)
- training set y (classes corresponding to features)
- test_instance (to be classified instance)
- k (determines how many neighbors are taken into account)
- weight (can be True / False. Determines if neighbors are weighted or not)

OUTPUT:
- class assignment for test instance
'''
def k_nearest_neighbors(training_set_x, training_set_y, test_instance, k, weight=False):    
    distances = []  
    result = 0
    # find all distances from test_instance to training_set
    for i in range(len(training_set_x)):
        distances.append((i, distance(training_set_x[i], test_instance)))
    # sort list: ascending distances
    distances.sort(key=operator.itemgetter(1))

    prediction = 0    
    # calculate non-weighed average
    if weight == False:
        for i in range(k):
            prediction += training_set_y[distances[:k][i][0]][0]
        
        result = round(prediction/k)
        
    # calculate weighed average
    if weight == True:
        # create weights out of normalized values for 1/distance
        weightlist = []
        for i in range(k):
            weightlist.append(1/distances[:k][i][1])
        normalizer = 1/sum(weightlist)        
        for i in range(k):
            weightlist[i] = weightlist[i] * normalizer
            
        # multiply with weight
        for i in range(k):
            prediction += weightlist[i] * training_set_y[distances[:k][i][0]][0]
        
        result = round(prediction)
                       
    return result
    
# example run without using weights
print k_nearest_neighbors(X_train[0], y_train, X_test[0], 5, weight=False)
# example run using weights
print k_nearest_neighbors(X_train, y_train, X_test[0], 5, weight=True)

1.0
1.0


### Question 1) Evaluation

In [4]:
'''
INPUT:
- training set features
- training set classes
- to be classified test set
- value for k
- weights (true or false)

OUTPUT:
- number of correctly classified instances
- number of incorrectly classified instances

In addition to returning values, the function also prettyprints the results
'''
def k_nn_evaluation(training_set_x, training_set_y, test_set_x, test_set_y, k, use_weight=False):    
    correct = 0
    false = 0

    for i in range(len(test_set_x)):       
        predicted = k_nearest_neighbors(training_set_x, training_set_y, test_set_x[i], k, use_weight)
        actual = test_set_y[i][0]      
        if predicted == actual:
            correct += 1
        if predicted != actual:
            false +=1
 
                
                
    accurracy = round((float(correct)/float(false + correct)), 5)
    
    print 'k =        ', k
    print 'Correct:   ', correct
    print 'False:     ', false
    print 'Accurracy: ', accurracy
    print ''
    
    return correct, false

##### Results for unweighted k-NN

In [5]:
# find accurracy for k = 1, 2, ..., 10   
total_correct = 0
total_false = 0
for i in range(10):        
    correct, false = k_nn_evaluation(X_train, y_train, X_test, y_test, i+1)
    total_correct += correct
    total_false += false
print 'Total correct:   ', total_correct
print 'Total false:     ', total_false
print 'Total accurracy: ', round((float(total_correct)/float(total_false+total_correct)), 5)

k =         1
Correct:    237
False:      3
Accurracy:  0.9875

k =         2
Correct:    238
False:      2
Accurracy:  0.99167

k =         3
Correct:    236
False:      4
Accurracy:  0.98333

k =         4
Correct:    238
False:      2
Accurracy:  0.99167

k =         5
Correct:    236
False:      4
Accurracy:  0.98333

k =         6
Correct:    235
False:      5
Accurracy:  0.97917

k =         7
Correct:    235
False:      5
Accurracy:  0.97917

k =         8
Correct:    235
False:      5
Accurracy:  0.97917

k =         9
Correct:    234
False:      6
Accurracy:  0.975

k =         10
Correct:    235
False:      5
Accurracy:  0.97917

Total correct:    2359
Total false:      41
Total accurracy:  0.98292


##### Results for weighted k-NN

In [6]:
# find accurracy for k = 1, 2, ..., 10   
total_correct = 0
total_false = 0
for i in range(10):        
    correct, false = k_nn_evaluation(X_train, y_train, X_test, y_test, i+1, use_weight=True)
    total_correct += correct
    total_false += false
print 'Total correct:   ', total_correct
print 'Total false:     ', total_false
print 'Total accurracy: ', round((float(total_correct)/float(total_false+total_correct)), 5)

k =         1
Correct:    237
False:      3
Accurracy:  0.9875

k =         2
Correct:    237
False:      3
Accurracy:  0.9875

k =         3
Correct:    236
False:      4
Accurracy:  0.98333

k =         4
Correct:    236
False:      4
Accurracy:  0.98333

k =         5
Correct:    236
False:      4
Accurracy:  0.98333

k =         6
Correct:    235
False:      5
Accurracy:  0.97917

k =         7
Correct:    235
False:      5
Accurracy:  0.97917

k =         8
Correct:    235
False:      5
Accurracy:  0.97917

k =         9
Correct:    234
False:      6
Accurracy:  0.975

k =         10
Correct:    235
False:      5
Accurracy:  0.97917

Total correct:    2356
Total false:      44
Total accurracy:  0.98167


### Discussion results weighted vs unweighted
The differences between weighted k-NN and unweighted k-NN are minimal. Accurracy for weighted equals 0.98167, while accurracy for unweighted equals 0.98292. We expected the weighted implementation of k-NN to perform better, because the calculation is more sophisticated. 



After inspecting the euclidean distances, we found that (almost) all distances are in the same range. Therefore, the assigned weights will be roughly the same. As a result we think adding the weights did not improve the algorithm.

### Question 1) C

In [7]:
# for every feature, make list. combine everything in one list.
single_features = []
for i in range(len(X_train[0])):    
    feature = []
    for j in range(len(X_train)):
        feature.append(X_train[j][i])
    single_features.append(feature)

In [8]:
# store accurracy of features in array, later use this as weight
feature_weights = np.array([])
for i in range(len(single_features)):
    print '--- feature ', i+1, ' ---'    
    correct, false = k_nn_evaluation(single_features[i], y_train, X_test, y_test, 4)    
    feature_weights = np.hstack([feature_weights, round((float(correct)/float(false+correct)), 5)])
    
# normalize weights
normalizer = 1/sum(feature_weights)
feature_weights *= normalizer

--- feature  1  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  2  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  3  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  4  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  5  ---
k =         4
Correct:    72
False:      168
Accurracy:  0.3

--- feature  6  ---
k =         4
Correct:    72
False:      168
Accurracy:  0.3

--- feature  7  ---
k =         4
Correct:    88
False:      152
Accurracy:  0.36667

--- feature  8  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  9  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  10  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  11  ---
k =         4
Correct:    80
False:      160
Accurracy:  0.33333

--- feature  12  ---
k =         4
Correct:    80
False:    

In [9]:
'''
Very similar to 'k_nearest_neighbors', but with an additional parameter for weights.
INPUT:
- training set X (training examples with features)
- training set y (classes corresponding to features)
- test_instance (to be classified instance)
- k (determines how many neighbors are taken into account)
- weights (vector with predictive values of features)

OUTPUT:
- class assignment for test instance
'''
def k_nearest_neighbors_featureweight(training_set_x, training_set_y, test_instance, k, weights):    
    distances = []  
    result = 0

    for i in range(len(training_set_x)):
        w_feature_distance = 0
        for j in range(len(weights)):
            w_feature_distance += distance(training_set_x[i] * weights, test_instance)            
        distances.append((i, w_feature_distance))

    # sort list: ascending distances
    distances.sort(key=operator.itemgetter(1))
    
    prediction = 0              
    for i in range(k):
        prediction += training_set_y[distances[:k][i][0]][0]
                               
    return round(prediction/k)

In [10]:
'''
INPUT:
- training set features
- training set classes
- to be classified test set
- value for k
- weights (vector with predictive values of features)

OUTPUT:
- number of correctly classified instances
- number of incorrectly classified instances

In addition to returning values, the function also prettyprints the results
'''
def k_nn_evaluation_featureweight(training_set_x, training_set_y, test_set_x, test_set_y, k, weights):    
    correct = 0
    false = 0
    
    for i in range(len(test_set_x)):       
        predicted = k_nearest_neighbors_featureweight(training_set_x, training_set_y, test_set_x[i], k, weights)
        actual = test_set_y[i][0]      
        if predicted == actual:
            correct += 1
        if predicted != actual:
            false +=1
                 
    accurracy = round((float(correct)/float(false + correct)), 5)
    
    print 'k =        ', k
    print 'Correct:   ', correct
    print 'False:     ', false
    print 'Accurracy: ', accurracy
    print ''
    
    return correct, false

In [17]:
# find accurracy for k = 1, 2, ..., 10   
total_correct = 0
total_false = 0
for i in range(10):        
    correct, false = k_nn_evaluation_featureweight(X_train, y_train, X_test, y_test, i+1, feature_weights)
    total_correct += correct
    total_false += false
print 'Total correct:   ', total_correct
print 'Total false:     ', total_false
print 'Total accurracy: ', round((float(total_correct)/float(total_false+total_correct)), 5)

KeyboardInterrupt: 

### Discussion results weighted features

Above results show that the weighing of features performs worse than using distance weight or no weight. Accurracy using weighted features is 79.7%, which is significantly worse than the other two implementations.

# Question 2

#### NN IMPLEMENTATION

In [23]:
def sigmoid(z):
    return 1/(1+np.exp(-z))

def sigmoid_derivative(z):
    return z*(1-z)

def forward(X, W, W_output):    
    steps = list()
    steps.append(X)    
    # loop over weights for hidden layers
    for i in range(len(W)):
        steps.append(sigmoid(np.dot(steps[i], W[i])))    
    # for final value, use W_output
    steps.append(sigmoid(np.dot(steps[-1], W_output)))
    
    return steps, steps[-1]

def weights(inputLayerSize, outputLayerSize, hiddenLayerSize, hiddenLayerDepth, value):    
    W = list()
    W_output = 0
    
    if value == -1:
        # create random weight values for all hidden layers
        W = list()
        W.append(np.random.randn(inputLayerSize, hiddenLayerSize))
        if hiddenLayerDepth > 1:
            for i in range(hiddenLayerDepth):
                W.append(np.random.randn(hiddenLayerSize, hiddenLayerSize))

        # W_output might have different dimensions 
        # (if outputLayerSize != hiddenLayerSize)
        W_output = np.random.randn(hiddenLayerSize, outputLayerSize)
    
    if value != -1:
        W = list()
        W.append(np.full((inputLayerSize, hiddenLayerSize), value))
        if hiddenLayerDepth > 1:
            for i in range(hiddenLayerDepth):
                W.append(np.full((hiddenLayerSize, hiddenLayerSize), value))
            
        # W_output might have different dimensions 
        # (if outputLayerSize != hiddenLayerSize)
        W_output = np.full((hiddenLayerSize, outputLayerSize), value)
        
    return W, W_output

In [19]:
def read_logistic_data(filename):
    my_data = genfromtxt(filename, delimiter=';')
    X = []
    y= []
    for i in range(len(my_data)):
        new_data = my_data[i]
        y.append([new_data[-1]])
        
        
        X.append(np.delete(new_data, -1))
        
        
        #X.append(np.delete(new_data, -1))
            
    return X, y

X, temp_y = read_logistic_data('digits123-1.csv')

'''
for this specific handwritten digit classification problem, we need to have 
three classes: 1, 2, 3. Representation of the classes will be as follows:

class '1' =   class '2' =   class '3' =
[1,           [0,           [0,
 0,            1,            0,
 0]            0]            1]
'''
y = []

for i in range(len(temp_y)):
    if temp_y[i][0] == 1.0:
        y.append(np.array([[1, 0, 0]]))        
    if temp_y[i][0] == 2.0:
        y.append(np.array([[0, 1, 0]]))      
    if temp_y[i][0] == 3.0:
        y.append(np.array([[0, 0, 1]]))

In [20]:
# chosen parameters for neural network:
inputLayerSize = 64   # 64 features
outputLayerSize = 3   # three classes: 1, 2, 3
hiddenLayerSize = 35  # simple rule of thumb: n of neurons in hidden layer
                      # is mean of inputLayerSize and outPutlayerSize.
hiddenLayerDepth = 1

# create weights for NN using above parameters
W, W_output = weights(inputLayerSize, outputLayerSize, hiddenLayerSize, hiddenLayerDepth, -1)

In [21]:
def backpropagation(X, y, W, W_output, alpha, iterations):
    count = 0    
    dy_dtheta1 = 0
    dy_dtheta2 = 0   
    while count < iterations:
        # loop over every training example. After loop, update weights.
        for i in range(len(X)):
            steps, resultY = forward(X[i], W, W_output)    

            # find delta_3: hypothesis - actual answer
            d_3 = resultY-y[i]       
            # find delta_2: ((output weights) . d_3 ) * g'(z(2))
            d_2 = np.dot(W_output, d_3[0].T) * sigmoid_derivative(steps[-2])
            
            # 3 outgoing arrows from hidden layer to final layer
            # [hidden] --> [hidden, hidden, hidden]
            # used for vectorization of code
            l2 = steps[-2].reshape(35,1)
            l2 = np.concatenate((l2, l2, l2), axis=1)        
                       
            # every input data point needs to connect to 35 hidden layers
            # [input] --> [input, input, ..., input] (width = 35 elements)
            l1 = steps[-3].reshape(64,1)
            for i in range(5):
                l1 = np.concatenate((l1, l1), axis = 1)    
            l1 = np.concatenate((l1, steps[-3].reshape(64,1), steps[-3].reshape(64,1), steps[-3].reshape(64,1)), axis=1)
            
            # add derivative of training example i
            dy_dtheta2 += l2*d_3
            dy_dtheta1 += l1*d_2

        # update weights
        W_updated = list()
        W_updated.append((1.0/len(X))*(W[0] - alpha * dy_dtheta1))
        W_output_updated = (1.0/len(X)) * (W_output - alpha * dy_dtheta2)
        
        W = W_updated
        W_output = W_output_updated
        
        count = count+1
  
    return W_updated, W_output_updated
        
    
W_resulting, W_output_resulting = backpropagation(X, y, W, W_output, 0.04, 500)  

In [22]:
def find_class(results):   
    
    if ((results[0] > results[1]) & (results[0] > results[2])):
        number = 1.0
    if ((results[1] > results[0]) & (results[1] > results[2])):
        number = 2.0
    if ((results[2] > results[0]) & (results[2] > results[1])):
        number = 3.0

    return number

#### Scikit-learn implementation

In [35]:
from sklearn import linear_model
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
#Dit moet je ff van te voren in je terminal runnen.
#pip install scikit-neuralnetwork
#pip install scikit-learn==0.18rc2

#Logistic regression
reg = linear_model.LinearRegression()

#Normal accuracy logistic regression
reg.fit(X_train, y_train)
print 'Accuracy logistic regression scikit: ', reg.score(X_test, y_test)

#Backpropagation
clf = MLPClassifier(alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=1, max_iter=1000)

clf.fit(X_train, y_train)

#K-nearest neighbour classifier
neigh = KNeighborsClassifier(n_neighbors=5)

#Normal accuracy neighbour
neigh.fit(X_train, y_train)
print 'Accuracy k-nearest scikit: ', neigh.score(X_test, y_test)

Accuracy logistic regression scikit:  0.780371206145
Accuracy k-nearest scikit:  0.979166666667


C:\Users\Yvonn\Anaconda2\lib\site-packages\ipykernel\__main__.py:24: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


### Test results

In [34]:


X_test, y_test = read_logistic_data('digits123-2.csv')
steps, resultY = forward(X_test, W_resulting, W_output_resulting)    


def correctness(resultY, y_test):
    correct = 0
    incorrect = 0
    for i in range(len(resultY)):
        if find_class(resultY[i]) == y_test[i][0]:
            correct += 1
        if find_class(resultY[i]) != y_test[i][0]:
            incorrect +=1
    return correct, incorrect

print 'Full test set digits123-2 used:'
correct, incorrect = correctness(resultY, y_test)
print 'n correct:   ', correct
print 'n incorrect: ', incorrect
print 'Accurracy our neural network:   ', float(correct)/(float(correct)+float(incorrect))
print 'Accuracy neural network scikit: ', clf.score(X_test, y_test)
print(len(X_test))

print 'Test set digits123-2 split in 2 for cross validation:'
def cross_validation(X_test, y_test):
    X_cross = X_test[:120]
    new_X_test = X_test[120:]
    y_cross = y_test[:120]
    new_y_test = y_test[120:]
    steps_cross, resultY_cross = forward(X_cross, W_resulting, W_output_resulting)
    steps_test, resultY_test = forward(new_X_test, W_resulting, W_output_resulting)
    correct_cross, incorrect_cross = correctness(resultY_cross, y_cross)
    correct_test, incorrect_test = correctness(resultY_test, new_y_test)
    print 'Accurracy our neural network cross set:   ', float(correct_cross)/(float(correct_cross)+float(incorrect_cross))
    print 'Accurracy our neural network test set:   ', float(correct_test)/(float(correct_test)+float(incorrect_test))
    print 'Accuracy neural network scikit cross set: ', clf.score(X_cross, y_cross)
    print 'Accuracy neural network scikit test set: ', clf.score(new_X_test, new_y_test)
cross_validation(X_test, y_test)   

Full test set digits123-2 used:
n correct:    228
n incorrect:  12
Accurracy our neural network:    0.95
Accuracy neural network scikit:  0.904166666667
240
Test set digits123-2 split in 2 for cross validation:
Accurracy our neural network cross set:    0.933333333333
Accurracy our neural network test set:    0.966666666667
Accuracy neural network scikit cross set:  0.916666666667
Accuracy neural network scikit test set:  0.891666666667


### Discussion question 2

Above results show the accuracy of our neural network and the neural network that is used by the scikit-learn package. It is shown that our accuracy is better. It is noticable that with cross validation that our neural network has a better accuracy in both sets. So we chose to try and optimize our own neural network even further for a better accuracy value.

First we tried to normalize our training and test set to try and retrieve better results. Sadly this was not the case, because the accuracy of our model decreased. After that we tried to use different weights for our neural network. Like 0.001, 0.01 and 0.1. This also resulted in a decrease in accuracy. After that we looked at the wrong predictions that were made by our neural network. It was noticable that the most errors that were made by our neural network were a wrong prediction of class 2. The neural network predicted that the digits belonged to class 1, when they actually belonged to class 2. We tried to introduce a bias for class 2, unfortunately this was unsuccesfull.

With this being said. We did not know how we could improve our neural network any further. So our final accuracy is 0.95. We are very pleased by this result.